### PR TITLE
Frontend i18n Refactor and Translation in JS

### PR DESCRIPTION
**Pull request by rtibbles**
## Summary

Refactors our i18n frontend code, and systematizes it into one module.

Adds capability to allow translations to happen in JS code, as discussed in https://github.com/learningequality/kolibri/issues/1144. Follows up on comments in there and deprecates `$trNamespace` in favour of the component `name`. Namespaces for JS files are freely determined.

Additionally, based on comments here: https://github.com/learningequality/kolibri/pull/1771#discussion_r127063224 have slightly refactored the way we handle text direction, and created a globally available vue property `isLtr` for easy checking.

It also adds translation for the page titles of the user profile plugin as a proof of concept.

Posting here as a WIP for feedback on API/approach.

## TODO

- [ ] Have tests been written for the new code?
- [x] Has documentation been written/updated?
- [x] Add an entry to CHANGELOG.rst

## Reviewer guidance

I went through and changed most extant `$trNamespace` to `name`, that commit can be ignored.

## Issues addressed

Lays the groundwork for and partly solves #1144.

### CODE REVIEW

File: frontend_build/src/extract_$trs.js [hunk 1]
```diff
@@ -19,7 +43,36 @@ extract$trs.prototype.apply = function(compiler) {
   compiler.plugin('emit', function(compilation, callback) {
     var messageExport = {};
     var nameSpaces = [];
-
+    function registerFoundMessages(messageNameSpace, messages, module) {
+      if (messageNameSpace) {
+        // Warn about duplicate nameSpaces *within* a bundle (no way to warn across).
+        if (nameSpaces.indexOf(messageNameSpace) !== -1) {
+          logging.error('Duplicate namespace ' + messageNameSpace + ' found in ' + module.resource);
```
Comment (christianmemije) [inline frontend_build/src/extract_$trs.js:50]: Update log error to use `name` instead of `namespace`

File: frontend_build/src/extract_$trs.js [hunk 2]
```diff
@@ -19,7 +43,36 @@ extract$trs.prototype.apply = function(compiler) {
...
+        } else if (Object.keys(messages).length) {
+          // Check that the trNameSpace id is camelCase.
+          if (!isCamelCase(messageNameSpace)) {
+            logging.error(
+              `Name or $trNameSpace id "${messageNameSpace}" should be in camelCase. Found in ${module.resource}`
```
Comment (christianmemije) [inline frontend_build/src/extract_$trs.js]: Update log error.

File: frontend_build/src/extract_$trs.js [hunk 3]
```diff
@@ -19,7 +43,36 @@ extract$trs.prototype.apply = function(compiler) {
...
     var messageExport = {};
     var nameSpaces = [];
-
+    function registerFoundMessages(messageNameSpace, messages, module) {
+      if (messageNameSpace) {
+        // Warn about duplicate nameSpaces *within* a bundle (no way to warn across).
+        if (nameSpaces.indexOf(messageNameSpace) !== -1) {
+          logging.error('Duplicate namespace ' + messageNameSpace + ' found in ' + module.resource);
+        } else if (Object.keys(messages).length) {
+          // Check that the trNameSpace id is camelCase.
```
Comment (christianmemije) [inline frontend_build/src/extract_$trs.js]: Remove mentions of `trNameSpace`

File: frontend_build/src/extract_$trs.js (hunk 1 above)
Comment (rtibbles) [reply frontend_build/src/extract_$trs.js:50]: It's still a namespace, it's just that we use the Vue `name` property for the namespace in .vue files.

File: kolibri/core/assets/src/utils/i18n.js [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+  zh: languageDensities.dense,
+};
+
+function setLanguageDensity(lang_code) {
+  const shortCode = lang_code.split('-')[0].toLowerCase();
+  // Set the exported languageDensity in JS
+  languageDensity = languageDensityMapping[shortCode] || languageDensities.englishLike;
+  // Set the body class for global typography
+  global.document.body.classList.add(`language-${languageDensity}`);
+}
```
Comment (indirectlylit) [inline kolibri/core/assets/src/utils/i18n.js:85]: awesome - didn't notice this before!

**Comment by rtibbles** [comment]
Now with dynamic language switching:

dynamiclanguageswitching

Feedback on flow here much appreciated - @jtamiace, @khangmach

**Comment by rtibbles** [comment]
Note: incomplete translations due to translations being incomplete :)

File: frontend_build/src/extract_$trs.js (hunk 3 above)
Comment (rtibbles) [reply frontend_build/src/extract_$trs.js]: Will do - need to get rid of it from the perseus renderer first!

**Comment by khangmach** [comment]
Selecting language using a dropdown is simple enough. I don't think this belongs in the app bar however, as it would make language selection too easy and prominent, isn't a means of navigation, and isn't a primary action. This should be embedded in a footer or settings page. But since we don't have either I would suggest placing it in the drawer as shown below. 

As many more of these facility agnostic settings pop up, I'm also thinking of alternatives like a more general "settings" page or having the facility page within Manage be "Settings" and then include BOTH the ability to adjust general settings (which may include language selection) and specific facility-related settings (what we have now).

**Comment by rtibbles** [comment]
I like the general overall approach - any thoughts on how we should present language selection on pages *without* a nav bar available (most common use case I can think of is the login screen).

**Comment by khangmach** [comment]
I could see language selection adjustable within login and onboarding as well. In the form of a selector on login, and for onboarding, designed more robustly like something below (including flags).

image

**Comment by rtibbles** [comment]
Although we are generally using two part locale codes that include country codes, I'm a little wary of using flags (even though they make a good visual indicator) for the reasons outlined in https://usersnap.com/blog/design-language-switch/ - one language for which this is particularly strong is Arabic, where there are lots of countries that speak Arabic, but far fewer locale codes that map to that language.

**Comment by khangmach** [comment]
Having a "Change language" option within the profile dropdown would also be acceptable IMO. See trello for example

**Comment by jtamiace** [comment]
I also think embedded in the app bar is an inappropriate place for the same reasons.

A settings page is ideal in the long run. If we do decide to put the switcher in the drawer I'd suggest placing it closer to the bottom of it.

^ Was just about to suggest putting it in the profile dropdown too, and activate a simple modal.

**Comment by rtibbles** [comment]
My only concern with hiding the language switcher in any way is the difficulty of finding it if the interface is in a language you cannot understand. I think the Facebook style language switcher is best for this kind of discoverability, but as @khangmach mentioned we do not currently have a footer.

Might it be worth adding a footer for this?

**Comment by jtamiace** [comment]
I guess there is also the issue of that language switcher becoming inaccessible to guest users if it's in the profile dropdown.

**Comment by rtibbles** [comment]
Right - worth thinking about this as a pivotal critical user pathway for all user types, I think.

**Comment by jtamiace** [comment]
Having a footer on the home page makes sense, where language setting can be added. But even for facebook, once the user logs in, changing language moves to the Settings page, which is accessible from a dropdown in the app bar.

screen shot 2017-07-14 at 9 33 26 am

**Comment by rtibbles** [comment]
That compromise seems reasonable to me - one way we can mitigate the issue of a user just 'walking away' from their session and leaving it in an incomprehensible state is to do client side idle checks. If a user has stepped away for 20 minutes, we can prompt if they want to stay logged in - if we don't get a response in 20 seconds, we log out and redirect to the login page?

(can do similarly for guests, but no need to logout)

**Comment by khangmach** [comment]
So are we doing both? Footer & profile dropdown with language selection modal?

**Comment by jtamiace** [comment]
Some ideas for location of language switcher across Facility users/guests:

GUESTS
Sign in page: footer 
Learn: Button (which opens a modal?) or dropdown in the drawer?

LOGGED IN FACILITY USER
Option in the profile dropdown

Another thought regarding concern about the user getting to the Learn interface and not knowing the language: even if the switcher is placed in the app bar, will they know that the dropdown represents switching the language? 

I also have concerns about making the language switcher too accessible to Learners in the classroom context. If Harish is concerned about the kids downloading a bunch of content, I can also imagine them switching the language for fun if it's accessible via the app bar or a universal footer when they are logged in.

**Comment by khangmach** [comment]
I don't think we should jump to an assessment that we need a footer. If we introduce a footer I think we should scope out it's purpose beyond just language select. So far the only things that would go in the footer are language selector and maybe the powered by kolibri text, which I don't think are enough to merit the development of one. 

I think having the language selector not as obvious to find (layered behind profile dropdown) works in our advantage for the reason Jessica mentioned above with kids playing with languages. Don't think we need a footer for login page. It could just exist in a corner with contrasting text

**Comment by rtibbles** [comment]
I am fine with the language selector being obscured on navigation pages (when combined with the 'timeout' behaviour I described above), but I do think it should be made very obvious on the login screen (and we should use a similar mechanism on the setup wizard too).

Whether that's a footer or a header, or a modal that pops up demanding the language first, I think it should be the first choice that a user makes (if they need to switch).

**Comment by jtamiace** [comment]
Agreed. What should the interaction for language switching be for Guest users that get to Learn? Would the drawer still be a good place...?

**Comment by rtibbles** [comment]
I think consistently keeping it in the Nav drawer works for all user types.

**Comment by indirectlylit** [comment]
I don't know if this has been mentioned - but for a first pass perhaps this could just be a facility configuration option?

Isn't it much more common that everyone in an organization will be using the same language?

**Comment by rtibbles** [comment]
Copied from a comment I made on slack:
> Current order of precedence:
> 1) Set by session by clicking the drop down
> 2) Set by cookie (currently we don't set a cookie, but could if we thought it was important to persist language across same browser logout/login)
> 3) Set by browser preferences (your browser sends a list of acceptable languages to the server and it chooses the highest priority one it can handle)
> 4) Set by server wide default setting

Could add items in places on this order (like a facility specific language setting for example)

**Comment by ivanistheone** [comment]
The [language selector next to the search](https://github.com/learningequality/kolibri/pull/1845#issuecomment-315257538) looks weird to me. 

Does it mean if I choose `LANG` I will only see search results with `language=LANG`? (seems too restrictive). Or will search look for all resources but prioritize those in language `LANG`? (I can see how that would be useful)

I would personally see the language dropdown as part of the profile dropdown, although a dropdown-within-a-dropdown is kind of horrible...

screen shot 2017-07-14 at 2 36 20 pm

**Comment by jtamiace** [comment]
Can't speak on the content part, but definitely don't want a dropdown in the dropdown... if it is in the profile dropdown, I'd imagine that we can present the language options with radio buttons in a pop up modal.

**Comment by jtamiace** [comment]
Alternatively in the modal, something structured like:

---

Current language: [Language]

Dropdown menu of language options

[modal buttons for save/cancel]

---

**Comment by rtibbles** [comment]
http://cdn3.meme.am/cache/instances/folder296/500x/78402296/xzibit-yo-dawg-i-heard-you-like-dropdowns-so-i-made-a-dropdown-and-put-a-dropdown-in-your-dropdown.jpg

**Comment by khangmach** [comment]
@ivanistheone no dropdowns within dropdowns. Here's a sample flow of how this could work, where clicking on change language prompts a simple modal/dialog to appear. Styles are for demonstration only

example

**Comment by indirectlylit** [comment]
The bulk of this work is ready and does not need additional design. I'd also like to potentially use it for some font-related work.

Could we move the "language switching" functionality to a new PR and continue the thread there?

File: kolibri/core/assets/src/utils/i18n.js [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+        },
+        error => {
+          logging.error(error);
+          logging.error('An error occurred trying to setup Internationalization', error);
+          reject();
+        }
+      );
+    } else if (global.languageCode === 'rt-lft') {
+      require.ensure([], () => {
+        toFakeRTL = require('./mirrorText').toFakeRTL;
```
Comment (indirectlylit) [inline kolibri/core/assets/src/utils/i18n.js:203]: fairly inconsequential, but it appears that `toFakeRTL` will not be set if the `Intl` polyfill is used

Comment (rtibbles) [reply kolibri/core/assets/src/utils/i18n.js:203]: Yeah, not sure this is a huge issue as this is not intended to be production code.

**Pull request merged**